File: Library/test/verify/aoj-1549_test.cpp
#define PROBLEM "http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1549"

#include "../../template/template.cpp"

#include "../../structure/wavelet/wavelet-matrix.cpp"

int main() {
  int n;
  cin >> n;

  const int OFS = 1e6;
  vector< int > as(n);
  for(int i = 0; i < n; i++) cin >> as[i], as[i] += OFS;
  CompressedWaveletMatrix< int, 17 > wm(as);

  int q;
  cin >> q;
  for(int i = 0; i < q; i++) {
    std::cout << "i: " << i << std::endl;
    int l, r, d;
    cin >> l >> r >> d;
    r++;
    d += OFS;
    int ans = OFS * 2;
    std::cout << "wm.rank(d, l): " << wm.rank(d, l) <<std::endl;
    std::cout << "wm.rank(d, r): " << wm.rank(d, r) <<std::endl;
    if(wm.rank(d, l) < wm.rank(d, r)) {
      ans = 0;
    } else {
      int succ = wm.next_value(l, r, d);
      std::cout << "succ: " << succ <<std::endl;
      if(~succ) chmin(ans, abs(succ - d));
      std::cout << "ans: " << ans <<std::endl;
      int pred = wm.prev_value(l, r, d);
      std::cout << "pred: " << pred <<std::endl;
      if(~pred) chmin(ans, abs(pred - d));
      std::cout << "ans: " << ans <<std::endl;
    }
    cout << ans << "\n";
  }
}

File: Library/test/verify/aoj-1549_test.py
# PROBLEM "http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1549"
# WaveletMatrix
from typing import *
try:
    import sys
    sys.path.insert(0, "/Users/tkrk/Document_local/Documents/kyopuro/Library/")
    from template import *
    from WaveletMatrix import *
except Exception as e:
    raise e

# import sys
# sys.stdout = open('output-py.txt', 'wt')


def main():
    n: int = int(input())
    OFS = int(1e6)
    a: List[int] = list(map(lambda x: int(x) + OFS, input().split()))

    wm: CompressedWaveletMatrix[int] = CompressedWaveletMatrix[int](
        (int, 17), a)

    q: int = int(input())
    for _ in range(q):
        # print(f"i: {_}")
        l, r, d = map(int, input().split())
        r += 1
        d += OFS
        ans: int = OFS * 2
        # print(f"wm.rank(d, l): {wm.rank(d, l)}")
        # print(f"wm.rank(d, r): {wm.rank(d, r)}")
        if(wm.rank(d, l) < wm.rank(d, r)):
            ans = 0
        else:
            succ: int = wm.next_value(l, r, d)
            # print(f"succ: {succ}")
            if(~succ):
                ans, _ = chmin(ans, abs(succ - d))
            # print(f"ans: {ans}")
            pred: int = wm.prev_value(l, r, d)
            # print(f"pred: {pred}")
            if(~pred):
                ans, _ = chmin(ans, abs(pred - d))
            # print(f"ans: {ans}")
        print(ans)


if __name__ == '__main__':
    main()
